`Promise` needs to implement `FromJSValConvertible`

It's needed by [ClipboardItem contructor](https://www.w3.org/TR/clipboard-apis/#dom-clipboarditem-clipboarditem)

```
error[E0277]: the trait bound `std::rc::Rc<<D as bindings::codegen::DomTypes::DomTypes>::Promise>: js::conversions::FromJSValConvertible` is not satisfied
   --> /servo/target/debug/build/script_bindings-1cfd0dc58b4959d3/out/Bindings/ClipboardItemBinding.rs:595:61
    |
595 |         let arg0: Record<DOMString, Rc<D::Promise>> = match FromJSValConvertible::from_jsval(*cx, HandleValue::from_raw(args.get(0)), ()) {
    |                                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `js::conversions::FromJSValConvertible` is not implemented for `std::rc::Rc<<D as bindings::codegen::DomTypes::DomTypes>::Promise>`
```

```
error[E0277]: the trait bound `promise::Promise: js::conversions::FromJSValConvertible` is not satisfied
   --> /servo/target/debug/build/script_bindings-1cfd0dc58b4959d3/out/DomTypeHolder.rs:320:20
    |
320 |     type Promise = crate::dom::promise::Promise;
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `js::conversions::FromJSValConvertible` is not implemented for `promise::Promise`
    |

note: required by a bound in `bindings::codegen::DomTypes::DomTypes::Promise`
   --> /servo/target/debug/build/script_bindings-1cfd0dc58b4959d3/out/DomTypes.rs:636:283
    |
636 | ...loc_size_of::MallocSizeOf + js::conversions::FromJSValConvertible;
    |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `DomTypes::Promise`
```
Needs more work. Other occurences of using dom object in Record are in testbinding and, correct me if I'm wrong, it works since `FromJSValConvertible` is implemented for `DomRoot<T>` which is not the case for `Rc<Promise>`.

--- a/components/script/dom/promise.rs
+++ b/components/script/dom/promise.rs
@@ -15,7 +15,7 @@ use std::ptr;
 use std::rc::Rc;
 
 use dom_struct::dom_struct;
-use js::conversions::ToJSValConvertible;
+use js::conversions::{ConversionResult, FromJSValConvertibleRc, ToJSValConvertible};
 use js::jsapi::{
     AddRawValueRoot, CallArgs, GetFunctionNativeReserved, Heap, JS_ClearPendingException,
     JS_GetFunctionObject, JS_NewFunction, JSAutoRealm, JSContext, JSObject,
@@ -403,5 +403,26 @@ impl PromiseHelpers<crate::DomTypeHolder> for Promise {
         value: impl ToJSValConvertible,
     ) -> Rc<Promise> {
         Promise::new_resolved(global, cx, value, CanGc::note())
+    }
+}
+
+impl FromJSValConvertibleRc for Promise {
+    #[allow(unsafe_code)]
+    unsafe fn from_jsval(
+        cx: *mut JSContext,
+        value: HandleValue,
+    ) -> Result<ConversionResult<Rc<Promise>>, ()> {
+        if value.get().is_null() {
+            return Ok(ConversionResult::Failure("null not allowed".into()));
+        }
+        if !value.get().is_object() {
+            return Ok(ConversionResult::Failure("not an object".into()));
+        }
+        rooted!(in(cx) let obj = value.get().to_object());
+        if !IsPromiseObject(obj.handle()) {
+            return Ok(ConversionResult::Failure("not a promise".into()));
+        }
+        let promise = Promise::new_with_js_promise(obj.handle(), SafeJSContext::from_ptr(cx));
+        Ok(ConversionResult::Success(promise))
     }
 }

--- a/components/script/dom/testbinding.rs
+++ b/components/script/dom/testbinding.rs
@@ -910,6 +910,7 @@ impl TestBindingMethods<crate::DomTypeHolder> for TestBinding {
     fn FuncControlledMethodDisabled(&self) {}
     fn FuncControlledMethodEnabled(&self) {}
 
+    fn PassRecordPromise(&self, _: Record<DOMString, Rc<Promise>>) {}
     fn PassRecord(&self, _: Record<DOMString, i32>) {}
     fn PassRecordWithUSVStringKey(&self, _: Record<USVString, i32>) {}
     fn PassRecordWithByteStringKey(&self, _: Record<ByteString, i32>) {}

--- a/components/script_bindings/codegen/Bindings.conf
+++ b/components/script_bindings/codegen/Bindings.conf
@@ -490,7 +490,7 @@ DOMInterfaces = {
 
 'Promise': {
     'spiderMonkeyInterface': True,
-    'additionalTraits': ["crate::dom::promise::PromiseHelpers<Self>"]
+    'additionalTraits': ["crate::dom::promise::PromiseHelpers<Self>", "js::conversions::FromJSValConvertibleRc"]
 },
 
 'Range': {

--- a/components/script_bindings/webidls/TestBinding.webidl
+++ b/components/script_bindings/webidls/TestBinding.webidl
@@ -479,6 +479,7 @@ interface TestBinding {
   sequence<sequence<long>> returnSequenceSequence();
   undefined passUnionSequenceSequence((long or sequence<sequence<long>>) seq);
 
+  undefined passRecordPromise(record<DOMString, Promise<undefined>> arg);
   undefined passRecord(record<DOMString, long> arg);
   undefined passRecordWithUSVStringKey(record<USVString, long> arg);
   undefined passRecordWithByteStringKey(record<ByteString, long> arg);

